src/ir/stack-utils.cpp

```diff
@@ -0,0 +1,272 @@
+
+StackSignature& StackSignature::operator+=(const StackSignature& next) {
+  assert(composes(next));
+  std::vector<Type> stack(results.begin(), results.end());
+  size_t required = next.params.size();
+  // Consume stack values according to next's parameters
+  if (stack.size() >= required) {
+    stack.resize(stack.size() - required);
+  } else {
+    if (!unreachable) {
+      // Prepend the unsatisfied params of `next` to the current params
+      size_t unsatisfied = required - stack.size();
+      std::vector<Type> newParams(next.params.begin(),
+                                  next.params.begin() + unsatisfied);
+      newParams.insert(newParams.end(), params.begin(), params.end());
+      params = Type(newParams);
+    }
+    stack.clear();
+  }
+  // Add stack values according to next's results
+  if (next.unreachable) {
+    results = next.results;
+    unreachable = true;
+  } else {
+    stack.insert(stack.end(), next.results.begin(), next.results.end());
+    results = Type(stack);
+  }
+  return *this;
+}
+
+StackSignature StackSignature::operator+(const StackSignature& next) const {
+  StackSignature sig = *this;
+  sig += next;
+  return sig;
+}
+
+StackFlow::StackFlow(Block* block) {
+  // Abstract out the logic for making a pass through the block
+  auto processBlock = [&block](auto process) {
+    for (auto* expr : block->list) {
+      process(expr, StackSignature(expr));
+    }
+    bool unreachable = block->type == Type::unreachable;
+    Type params = unreachable ? Type::none : block->type;
+    process(block, StackSignature(params, Type::none, unreachable));
+  };
+
+  // We need to make an initial pass through the block to figure out how many
+  // values each unreachable instruction produces.
+  std::unordered_map<Expression*, size_t> producedByUnreachable;
+  {
+    size_t stackSize = 0;
+    size_t produced = 0;
+    Expression* lastUnreachable = nullptr;
+    processBlock([&](Expression* expr, StackSignature sig) {
+      // Consume params
+      if (sig.params.size() > stackSize) {
+        // We need more values than are available, so they must come from the
+        // last unreachable.
+        assert(lastUnreachable);
+        produced += sig.params.size() - stackSize;
+        stackSize = 0;
+      } else {
+        stackSize -= sig.params.size();
+      }
+
+      // Handle unreachable or produce results
+      if (sig.unreachable) {
+        if (lastUnreachable) {
+          producedByUnreachable[lastUnreachable] = produced;
+          produced = 0;
+        }
+        assert(produced == 0);
+        lastUnreachable = expr;
+        stackSize = 0;
+      } else {
+        stackSize += sig.results.size();
+      }
+    });
+
+    // Finish record for final unreachable
+    if (lastUnreachable) {
+      producedByUnreachable[lastUnreachable] = produced;
+    }
+  }
+
+  // Take another pass through the block, recording srcs and dests.
+  std::vector<Location> values;
+  Expression* lastUnreachable = nullptr;
```

Comment:
For this and also the `processBlock` call above, I can understand how these status-keeping variables work between instructions within a block, but do these work in the same way when we go from the last instruction of a block to the block itself?

I also find the abstraction of `processBlock` is little hard to understand, because basically it does two things: iterate children and process the block itself. I'm not sure, but could putting the content of `processBlock` in the function main body and factoring out separate tasks (1. populating `producedByUnreachable` 2. populating src and dst locations) as separate functions possibly make this more readable?

Reply:
The goal of `processBlock` is to encapsulate the logic necessary to treat the block and its children uniformly. Treating the block and its children uniformly is important to avoid having to duplicate the entire bodies of the two tasks, so the tasks definitely need to be factored into lambdas like they are now. Since the tasks are already in lambdas, it seemed better to extract the processing structure into its own lambda as well rather than duplicating it. Hopefully this will make more sense to the reader once I explain in the code how `processBlock` manages to treat the block and its instructions uniformly.